Add circuit breaking task (#1597)
linkerd/website: linkerd.io/content/2-edge/tasks/circuit-breakers.md

+++
title = "Circuit Breakers"
description = "Protect against service failures using circuit breakers"
+++

Circuit breaking is a powerful feature where Linkerd will temporarily stop

> routing requests to an endpoint if that endpoint is deemed to be unhealthy;

Reviewer: i _think_ a comma is more correct here?
```suggestion
routing requests to an endpoint if that endpoint is deemed to be unhealthy,
```

instead routing that request to other replicas in the service.

> In this tutoral, we'll see how to enable circuit breaking on a service to

Reviewer: should we capitalize "Service" here to make it clear that we are referring specifically to the Kubernetes Service object? might make it clearer that you can't put circuit breaking annotations on random stuff...

improve client success rate when a backend replica is unhealthy.

{{< trylpt >}}

## Prerequisites

To use this guide, you'll need a Kubernetes cluster running:

- Linkerd and Linkerd-Viz. If you haven't installed these yet, follow the
  [Installing Linkerd Guide](../install/).

## Set up the demo

> Remember those puzzles where one guard always tells the truth and one guard

Reviewer: love this :)

kind: Service
metadata:
  name: bb
  namespace: circuit-breaking-demo
spec:
  ports:
  - name: http
    port: 8080
    targetPort: 8080
  selector:
    app: bb
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: slow-cooker
  namespace: circuit-breaking-demo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: slow-cooker
  template:
    metadata:
      labels:
        app: slow-cooker
    spec:
      containers:
      - args:
        - -c
        - |
          sleep 5 # wait for pods to start
          /slow_cooker/slow_cooker --qps 10 http://bb:8080
        command:
        - /bin/sh
        image: buoyantio/slow_cooker:1.3.0
        name: slow-cooker
EOF
```

We can now look at the success rate of the `good` and `bad` pods:

```console
> linkerd viz -n circuit-breaking-demo stat deploy
NAME          MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
bad              1/1     6.43%   4.7rps           1ms           1ms           4ms          2
good             1/1   100.00%   5.9rps           1ms           1ms           1ms          3
slow-cooker      1/1   100.00%   0.3rps           1ms           1ms           1ms          1
```

> Here we can see that `good` and `bad` deployments are each reciving similar

Reviewer: typo:
```suggestion
Here we can see that `good` and `bad` deployments are each receiving similar
```

> amounts of traffic but `good` has a success rate of 100% while the success

Reviewer: ```suggestion
amounts of traffic, but `good` has a success rate of 100% while the success
```

rate of `bad` is very low (only healthcheck probes are succeeding). We can also
see how this looks from the perspective of the traffic generator:

```console
> linkerd viz -n circuit-breaking-demo stat deploy/slow-cooker --to svc/bb
NAME          MESHED   SUCCESS       RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
slow-cooker      1/1    51.00%   10.0rps           1ms           1ms           2ms          2
```

From `slow-cooker`'s perspective, roughly 50% of requests that it sends to the
service are failing. We can use circuit breaking to improve this by cutting off
traffic to the `bad` pod.

## Breaking the circuit

> Linkerd supports a type of circuit breaking called _consecutive failure accrual_.

Reviewer: maybe worth linking to the reference docs here when they merge?

Reviewer 2: Definitely.

This works by tracking consecutive failures from each endpoint in Linkerd's
internal load balancer. If there are ever too many failures in a row, that

> endpoint is temporarily ignored and Linkered will only load balance among the

Reviewer: typo:
```suggestion
endpoint is temporarily ignored and Linkerd will only load balance among the
```

{
  "Kind": {
    "ConsecutiveFailures": {
      "max_failures": 7,
      "backoff": {
        "min_backoff": {
          "seconds": 1
        },
        "max_backoff": {
          "seconds": 60
        },
        "jitter_ratio": 0.5
      }
    }
  }
}
```

This tells us that Linkerd will use `ConsecutiveFailures` failure accrual
when talking to the `bb` service. It also tells us that the `max_failures` is
7, meaning that it will trip the circuit breaker once it observes 7 consective
failures. We'll talk more about each of the parameters here at the end of this
article.

Let's look at how much traffic each pod is getting now that the circuit breaker
is in place:

```console
> linkerd viz -n circuit-breaking-demo stat deploy
NAME          MESHED   SUCCESS       RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
bad              1/1    94.74%    0.3rps           1ms           1ms           1ms          3
good             1/1   100.00%   10.3rps           1ms           1ms           4ms          4
slow-cooker      1/1   100.00%    0.3rps           1ms           1ms           1ms          1
```

Notice that the `bad` pod's RPS is significantly lower now. The circuit breaker

> has stoped nearly all of the traffic from `slow-cooker` to `bad`.

Reviewer: typo:
```suggestion
has stopped nearly all of the traffic from `slow-cooker` to `bad`.
```

Reviewer: should we maybe mention that the reason it's "nearly all" is due to probes?

Author: it's also probation requests (or whatever we call them) and I didn't really want to get into all that here.

Reviewer: that's the kind of probe i meant, actually. should maybe call them "probation requests" instead of "probe requests"

Author: oh, well it's ALSO kubernetes liveness probes 😅

We can also see how this has affected `slow-cooker`:

```console
> linkerd viz -n circuit-breaking-demo stat deploy/slow-cooker --to svc/bb
NAME          MESHED   SUCCESS       RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
slow-cooker      1/1    99.83%   10.0rps           1ms           1ms           1ms          4
```

> Nearly all of `slow-cookers` requests are now getting routed to the `good` pod

Reviewer: typo: 
```suggestion
Nearly all of `slow-cooker`'s requests are now getting routed to the `good` pod
```

and succeeding!

## Tuning circuit breaking

As we saw when we ran the `linkerd diagnostics policy` command, consecutive
failure accrual is controlled by a number of parameters. Each of these

> parameters has a default, but can manually configured using annotations:

Reviewer: ```suggestion
parameters has a default, but can be manually configured using annotations:
```

    setting a higher value if you find circuit breakers are tripping too easily,
    causing traffic to be cut off from healthy endpoints.
- `balancer.linkerd.io/failure-accrual-consecutive-max-penalty`
  - The maximum amount of time a circuit breaker will remain tripped
    before the endpoint is restored (default: 60s). Consider setting a longer
    duration if you want to reduce the amount of traffic to endpoints which have
    tripped the circuit breaker. Consider setting a shorter duration if you'd
    like tripped circuit breakers to recover faster after an endpoint becomes
    healthy again.
- `balancer.linkerd.io/failure-accrual-consecutive-min-penalty`
  - The minimum amount of time a circuit breaker will remain tripped
    before the endpoints is restored (default: 1s). Consider tuning this in a
    similar way to `failure-accrual-consecutive-max-penalty`.
- `balancer.linkerd.io/failure-accrual-consecutive-jitter-ratio`
  - The amount of jitter to introduce to circuit breaker backoffs (default: 0.5).
    You are unlikely to need to tune this but might consider increasing it if
    you notice many clients are sending requests to a circuit broken endpoint

>     at the same time, leading to spikey traffic patterns.

Reviewer: [had to look in the dictionary](https://www.dictionary.com/browse/spiky) but i think it's "spiky" rather than "spikey"
```suggestion
    at the same time, leading to spiky traffic patterns.
```
